Firestore/core/src/remote/serializer.h

```diff
@@ -171,7 +171,7 @@ class Serializer {
 
   static google_firestore_v1_DocumentMask EncodeFieldMask(
       const model::FieldMask& mask);
-  static model::FieldMask DecodeFieldMask(
+  static util::StatusOr<model::FieldMask> DecodeFieldMask(
```

Comment:
This method should take a `Reader` and should just return the `FieldMask`. This will significantly reduce the burden on all the callers.

This will also get rid of all the early returns you've added in the implementation. Once the reader has failed, the result of these values doesn't matter so you'll be able to get rid of the `InvalidOrderBy` method and similar--the result is already going to be invalid.

Reply:
Much cleaner. Thanks.